### PR TITLE
ARROW-1689: [Python] Allow user to request no data copies

### PR DESCRIPTION
**Pull request by njwhite**
This makes performance debugging much easier, as it allows you to track down what (Arrow) data is causing unexpected delays in loading. It also makes testing features like ARROW-1689 easier as you can prove (via unit tests) that copies are not being made.

### CODE REVIEW

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -159,7 +159,7 @@ cdef class Column:
         sp_column.reset(new CColumn(boxed_field.sp_field, arr.sp_array))
         return pyarrow_wrap_column(sp_column)
 
-    def to_pandas(self, strings_to_categorical=False):
+    def to_pandas(self, strings_to_categorical=False, zero_copy_only=False):
```
Comment (xhochy) [inline python/pyarrow/table.pxi]: I would guess that this nearly always fails to the BlockManager. It will work for Arrays/Series but as Pandas will always allocate a large matrix for columns of the same type, you most likely get near to no zero-copies. (e.g. a DataFrame with two float64 columns will need a copy.)

**Review by xhochy — COMMENTED** [review]
I really like the idea of introducing such a flag. I guess due to Pandas nature the `to_pandas` will mostly have copies inside. I'm really looking forward to having this option on the inverese (`form_pandas`) in addition to a flag `nan_is_null` which can be set to `False`.

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (njwhite) [reply python/pyarrow/table.pxi]: @xhochy That makes sense - but why wouldn't you just make one `PandasBlock` per column (without copying), instead of created a `PandasBlock` per type and copying the various columns of that type into it? 

Also, I've pushed a fix for the AppVeyor warning but the Travis errors seem unrelated (in the Go & JS codebases)?

Comment (xhochy) [reply python/pyarrow/table.pxi]: @njwhite Because that is the underlying assumption based upon which Pandas 0.x DataFrames work on. There are several functions that use this assumptions to provide certain (slicing) features. @wesm might be able to go into detail/be more concrete.

**Comment by wesm** [comment]
pandas aggressively consolidates blocks. For non-categorical types, the maximum performance, most memory efficient choice for nearly all pandas users is to create pre-consolidated blocks. 

I’m a bit worried about doing significant work on this code without having ASV benchmarks set up

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (njwhite) [reply python/pyarrow/table.pxi]: @xhochy Makes sense, definitely want to keep this change focussed so should I just remove the flag from `pa.Table`? I'll fix the AppVeyor complaints tonight, & rebase off master to see if that fixes the other issues.

Comment (xhochy) [reply python/pyarrow/table.pxi]: No, keep it. I just wanted to give you the hint that it will likely fail to do zero-copy in >90% of all cases.

**Comment by njwhite** [comment]
@xhochy OK - Travis, &c. are all green now. @wesm thanks - that makes sense. I'm definitely not looking to do significant work! That said, [MAP_FIXED](https://stackoverflow.com/questions/10454964/mapping-non-contiguous-blocks-from-a-file-into-contiguous-memory-addresses) looks like a path to mapping different columns in the Arrow file to a contiguous address space (i.e. a pre-consolidated block) without physically copying the data...:)

File: cpp/src/arrow/status.h [hunk 2]
```diff
@@ -95,7 +95,8 @@ enum class StatusCode : char {
   PythonError = 12,
   PlasmaObjectExists = 20,
   PlasmaObjectNonexistent = 21,
-  PlasmaStoreFull = 22
+  PlasmaStoreFull = 22,
+  CopyRequired = 23
```
Comment (wesm) [inline cpp/src/arrow/status.h]: I don't think a new status code is needed. Can you return `Invalid` instead? As a matter of style, we should try not to use Status for routine error handling in C++. I think having this bubble up as an exception in Python is OK for now, but if we need to do zero-copy detection in C++, we are going to want to handle that in a different way.

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 3]
```diff
@@ -1542,26 +1566,19 @@ class ArrowDeserializer {
...
 
-    PyObject* dictionary;
-
-    // Release GIL before calling ConvertArrayToPandas, will be reacquired
-    // there if needed
-    lock.release();
-    RETURN_NOT_OK(
-        ConvertArrayToPandas(options_, dict_type->dictionary(), nullptr, &dictionary));
-    lock.acquire();
-
```
Comment (wesm) [inline cpp/src/arrow/python/arrow_to_pandas.cc]: Why is this deleted?

File: cpp/src/arrow/python/python-test.cc [hunk 4]
```diff
@@ -86,7 +86,7 @@ TEST(PandasConversionTest, TestObjectBlockWriteFails) {
 
   PyObject* out;
   Py_BEGIN_ALLOW_THREADS;
-  PandasOptions options;
+  PandasOptions options = {false, false};
```
Comment (wesm) [inline cpp/src/arrow/python/python-test.cc]: Can you instead add a default ctor to `PandasOptions`?

File: python/pyarrow/tests/test_convert_pandas.py [hunk 5]
```diff
@@ -212,6 +212,18 @@ def test_float_no_nulls(self):
...
+        result = pa.array([0, 1, 2]).to_pandas(zero_copy_only=True)
+        npt.assert_array_equal(result, [0, 1, 2])
+
+    def test_zero_copy_failure_on_object_types(self):
+        with self.assertRaises(pa.ArrowException):
+            pa.array(['A', 'B', 'C']).to_pandas(zero_copy_only=True)
+
+    def test_zero_copy_failure_when_nulls(self):
+        with self.assertRaises(pa.ArrowException):
+            pa.array([0, 1, None]).to_pandas(zero_copy_only=True)
```
Comment (wesm) [inline python/pyarrow/tests/test_convert_pandas.py]: There are 7 places where a zero copy error is being returned. Let's make sure we have unit tests that hit each of them

**Review by wesm — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/python/arrow_to_pandas.cc (hunk 3 above)
Comment (njwhite) [reply cpp/src/arrow/python/arrow_to_pandas.cc]: It's already been run by [this call](https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/arrow_to_pandas.cc#L1546) to Write [here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/arrow_to_pandas.cc#L1017-L1020) - so this change just reuses the save dictionary instead of building it again.

File: python/pyarrow/tests/test_convert_pandas.py (hunk 5 above)
Comment (njwhite) [reply python/pyarrow/tests/test_convert_pandas.py]: Added!

File: cpp/src/arrow/status.h (hunk 2 above)
Comment (njwhite) [reply cpp/src/arrow/status.h]: Done!

File: cpp/src/arrow/python/python-test.cc (hunk 4 above)
Comment (njwhite) [reply cpp/src/arrow/python/python-test.cc]: Done!

**Review by wesm — APPROVED** [review]
+1, thanks @njwhite!
